gemonlinedb/test/testXMLSerialization.cc

```diff
@@ -0,0 +1,182 @@
+
+    auto root = dom->getDocumentElement();
+
+    BOOST_CHECK(root != nullptr);
+    BOOST_CHECK(root->getChildNodes()->getLength() == 2);
+
+    BOOST_CHECK(detail::transcode(root->getFirstChild()->getNodeName()) ==
+                std::string("HEADER"));
+    BOOST_CHECK(detail::transcode(root->getLastChild()->getNodeName()) ==
+                std::string("DATA_SET"));
+}
+
+class AlwaysThrowErrorHandler : public xercesc::ErrorHandler
+{
+public:
+    virtual void warning(const xercesc::SAXParseException &e) { throw e; }
+    virtual void error(const xercesc::SAXParseException &e) { throw e; }
+    virtual void fatalError(const xercesc::SAXParseException &e) { throw e; }
+    virtual void resetErrors() {}
+};
+
+BOOST_AUTO_TEST_CASE(MakeDOMXsdValidation)
+{
+    XERCES_CPP_NAMESPACE_USE
+
+    auto builder = createTestXMLSerializationData();
+    auto dom = builder.makeDOM();
+
+    try {
+        // In principle, DOM 3 has an API to validate documents. One should
+        // modify the configuration of the document (getDOMConfig) and call
+        // normalizeDocument().
+        // However, the DOMImplementation available in Xerces doesn't implement
+        // this particular feature (an exception is thrown when trying to use
+        // it). Documents can, however, be validated at load time.
+        // So below we serialize the document to a memory buffer and validate it
+        // by reading it back.
+
+        // Get an implementation
+        auto impl = DOMImplementationRegistry::getDOMImplementation("LS"_xml);
+        BOOST_REQUIRE(impl != nullptr);
+
+        auto implLS = dynamic_cast<DOMImplementationLS *>(impl);
+
+        // Create the serializer, write and delete it
+        auto lsser = implLS->createLSSerializer();
+        BOOST_REQUIRE(lsser != nullptr);
+
+        // Serialize
+        auto buffer = lsser->writeToString(dom->getDocumentElement());
+        BOOST_REQUIRE(buffer != nullptr);
+        BOOST_REQUIRE(XMLString::stringLen(buffer) > 0);
+
+        // Create an input source and a parser
+        MemBufInputSource source(
+            reinterpret_cast<XMLByte *>(buffer),
+            XMLString::stringLen(buffer) * 2, // * 2 because of UTF-16 encoding
+            "fake"_xml, // Doc says it's used as a "fake system id"
+            true); // The buffer will be freed automatically
+        source.setEncoding("UTF-16"_xml);
+
+        auto errorHandler = std::make_shared<AlwaysThrowErrorHandler>();
+
+        XercesDOMParser parser;
+        parser.setErrorHandler(errorHandler.get());
+        parser.setExternalNoNamespaceSchemaLocation("schema/VFAT3ChipConfiguration.xsd");
+        parser.setValidationScheme(XercesDOMParser::Val_Always);
+        parser.setDoNamespaces(true);
+        parser.setDoSchema(true);
+        parser.setValidationSchemaFullChecking(true);
+
+        // Validate
+        parser.parse(source);
+
+        // Cleanup
+        delete lsser;
+        delete implLS;
```

Comment:
You shouldn't `delete` anything you didn't allocate, i.e., create with `new`.
Perhaps behind the scenes `xercesc` didn't do some cleanup correctly so it leaks here, but it's potentially problematic in that there *could* be a double free/delete problem down the road.